woocommerce/woocommerce-ios
# Bundle products in order form: Networking & Yosemite layer changes

Part of #10428 


⚠️ Please make sure https://github.com/woocommerce/woocommerce-ios/pull/10826 is approved before reviewing this PR

## Why


In https://github.com/woocommerce/woocommerce-ios/pull/10826, the new attributes/relationship were added to `ProductBundleItem` in storage. This PR adds the equivalent properties to `ProductBundleItem` in the Networking layer (readonly version) with the storage <--> readonly conversion, plus the logic on how bundle configuration is encoded when creating/updating `OrderItem` whe the order item is a bundle product.

## How

### Order item's bundle configuration

In order to configure a bundle product in an order item remotely, a `bundle_configuration` field is passed to the order line item with properties in the [API doc](https://woocommerce.com/document/bundles-rest-api-reference/#bundle-configuration-properties). This field is only passed when creating/updating an order, and does not exist in the API response. To represent each configuration item (bundled item in a bundle order item), `OrderItemBundleItem` struct was created that conforms only to `Encodable`. This `bundle_configuration` field is only encoded in an order line item when the array is non-empty.

### Variable product's bundle items

On the other hand, to display the form for configuring an order item variable product's bundled item, the new properties were added to the pre-existing `ProductBundleItem` that corresponds to the schema updates in https://github.com/woocommerce/woocommerce-ios/pull/10826.

## Testing instructions


A confidence check would be great on the order form to ensure the PR changes do not affect the current app behavior:

- Go to the Orders tab
- Tap on the `+` to create an order
- In the order creation form, add at least a product as an order item
- Tap `Create` to create the order remotely --> the order should be created successfully with the selected product(s)
- Go back to the Orders tab
- Tap on the previously created order
- Tap on `Edit` to edit the order
- Add at least another product as an order item --> the order should be updated successfully with the selected product(s)

---
